### PR TITLE
Add recent files

### LINKED ISSUE #239: remember recently opened files
It would be helpful to have a list of recently opened files to switch between `todo.txt` and `done.txt`. Since the application doesn't even remember the last used folder it's very fiddle to switch now. This could be unnecessary when implementing #144 if the session is remembered during restart of the application.

### PR DESCRIPTION
**Pull request by PKEv**
add recent files list in menu File
fix #239

### CODE REVIEW

**Comment by PKEv** [comment]
find bug.... :-(

**Comment by PKEv** [comment]
ok

**Comment by PKEv** [comment]
@oroulet need freback

File: qtodotxt/ui/controllers/main_controller.py [hunk 1]
```diff
@@ -387,6 +388,23 @@ def openFileByName(self, filename):
...
         logger.debug('Adding {} to watchlist'.format(filename))
         self._fileObserver.addPath(self._file.filename)
+        self.updateRecentFile()
+
+    def updateRecentFile(self):
+        lastOpenedArray = self._menu_controller.getRecentFileNames()
+        if self._file.filename in lastOpenedArray:
+            pos = lastOpenedArray.index(self._file.filename)
+            del lastOpenedArray[pos]
+            lastOpenedArray.insert(0, self._file.filename)
```
Comment (oroulet) [inline qtodotxt/ui/controllers/main_controller.py]: you can simply write:
muylist.remove(mystring)
mylist.insert(0, mystring)

File: qtodotxt/ui/controllers/menu_controller.py [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
 from qtodotxt.ui.dialogs import about_dialog
 from qtodotxt.ui.resource_manager import getIcon
 from qtodotxt.ui.dialogs.settings import Settings
 
 
 class MenuController(QtCore.QObject):
+
+    maxRecentFiles = 3
+    recentFileArray = []
+
```
Comment (oroulet) [inline qtodotxt/ui/controllers/menu_controller.py]: no need to use class variable. use object vairables

**Comment by oroulet** [comment]
The UI looks OK. But you should save the recent files in config files, shouldn't you? looks like you only keep things in memory

File: qtodotxt/ui/controllers/main_controller.py [hunk 3]
```diff
@@ -387,6 +388,23 @@ def openFileByName(self, filename):
...
+        else:
+            lastOpenedArray.pop()
+            lastOpenedArray.insert(0, self._file.filename)
+        ind = 1
+        for file in lastOpenedArray:
+            name = "lastOpened" + str(ind)
+            self._settings.setValue(name, file)
```
Comment (PKEv) [inline qtodotxt/ui/controllers/main_controller.py]: here i saved all recent files

**Comment by oroulet** [comment]
Why not save a list? That is the usual way to do that kind of things.
SetValue('recent_files', my list[:max_count])

File: qtodotxt/ui/controllers/menu_controller.py (hunk 2 above)
Comment (PKEv) [inline qtodotxt/ui/controllers/menu_controller.py]: I understand what you mean...

**Comment by PKEv** [comment]
it's strange but when I save a variable, then reading back I get it with additional characters, while at the same time it is stored in the registry properly
i save list[], but then read i got list[[]]....

File: qtodotxt/ui/controllers/main_controller.py [hunk 4]
```diff
@@ -387,6 +388,17 @@ def openFileByName(self, filename):
         self._settings.sync()
         logger.debug('Adding {} to watchlist'.format(filename))
         self._fileObserver.addPath(self._file.filename)
+        self.updateRecentFile()
+
+    def updateRecentFile(self):
+        lastOpenedArray = self._menu_controller.getRecentFileNames()
+        if self._file.filename in lastOpenedArray:
+            lastOpenedArray.remove(self._file.filename)
+        if len(lastOpenedArray) > self._menu_controller.maxRecentFiles:
```
Comment (oroulet) [inline qtodotxt/ui/controllers/main_controller.py]: lastopenedArray = lastopenedArray[:self._menu_controller.maxRecentFiles]

File: qtodotxt/ui/controllers/menu_controller.py [hunk 5]
```diff
@@ -23,13 +27,50 @@ def _initFileMenu(self):
...
         fileMenu.addAction(self._createPreferenceAction())
         fileMenu.addSeparator()
         fileMenu.addAction(self._createExitAction())
 
+    def updateRecentFileActions(self):
+        recentFileNames = self.getRecentFileNames()
+        ind = 1
+        for i in range(len(recentFileNames)):
+            try:
+                exist = os.path.isfile(recentFileNames[i])
```
Comment (oroulet) [inline qtodotxt/ui/controllers/menu_controller.py]: do we really need to check if file exist? A user may want to put files at the same place but sometime not mount netwoprk drive or whatever...

File: qtodotxt/ui/controllers/menu_controller.py [hunk 6]
```diff
@@ -23,13 +27,50 @@ def _initFileMenu(self):
...
+            self.recentFileArray[i].setVisible(True)
+            self.recentFileArray[i].setShortcuts(["Ctrl+" + str(ind)])
+            ind += 1
+
+    def getRecentFileNames(self):
+        recentFileNames = []
+        recentFileNames.append(QtCore.QSettings().value("lastOpened", []))
+        recentFileNamesStr = []
+        for name in recentFileNames[0]:
+            recentFileNamesStr.append(name)
```
Comment (oroulet) [inline qtodotxt/ui/controllers/menu_controller.py]: recentFileNamesStr = [name for name in recentFileNames[0]]

but why not return recentFileNames[0] directly?

File: qtodotxt/ui/controllers/menu_controller.py [hunk 7]
```diff
@@ -23,13 +26,43 @@ def _initFileMenu(self):
...
         fileMenu.addSeparator()
         fileMenu.addAction(self._createExitAction())
 
+    def updateRecentFileActions(self):
+        recentFileNames = self.getRecentFileNames()
+        ind = 1
+        for i in range(len(recentFileNames)):
+            if recentFileNames[i] is None:
+                self.recentFileArray[i].setVisible(False)
+                continue
```
Comment (oroulet) [inline qtodotxt/ui/controllers/menu_controller.py]: but this probably the last comment ;-). It was a good addition

Comment (PKEv) [inline qtodotxt/ui/controllers/menu_controller.py]: You're right - it is a strange code

**Pull request merged**